### PR TITLE
add experimental all_settled method to Mojo::Promise

### PR DESCRIPTION
**Pull request by elcamlost**
### Summary
Add experimental all_settled method to Mojo::Promise

### Motivation
Promise.allSettled is already stage 4 proposal and soon will be included at EcmaScript standard. It is also already supported in Chrome and soon will be supported in Firefox.  

### References
Closes https://github.com/mojolicious/mojo/issues/1423
Original TC39 Proposal https://github.com/tc39/proposal-promise-allSettled

### CODE REVIEW

File: t/mojo/promise.t [hunk 1]
```diff
@@ -213,6 +213,64 @@ Mojo::IOLoop->one_tick;
...
+$promise3->reject('third');
+$promise->reject('first');
+Mojo::IOLoop->one_tick;
+is_deeply \@results,
+  [
+  {status => 'rejected', reason => ['second']},
+  {status => 'rejected', reason => ['first']},
+  {status => 'rejected', reason => ['third']},
+  ],
+  'promises resolved';
```
Comment (kraih) [inline t/mojo/promise.t]: They are not.

File: t/mojo/promise.t [hunk 2]
```diff
@@ -213,6 +213,64 @@ Mojo::IOLoop->one_tick;
...
+# All Settled some rejected
+$promise  = Mojo::Promise->new->then(sub {@_});
+$promise2 = Mojo::Promise->new->then(sub {@_});
+$promise3 = Mojo::Promise->new->then(sub {@_});
+@results  = ();
+Mojo::Promise->all_settled($promise2, $promise, $promise3)
+  ->then(sub { @results = @_ });
+$promise2->reject('second');
+$promise3->resolve('third');
+$promise->resolve('first');
```
Comment (kraih) [inline t/mojo/promise.t]: These tests are not good, they don't actually ensure that the timing was right.

File: lib/Mojo/Promise.pm [hunk 3]
```diff
@@ -288,6 +310,27 @@ rejects. If the returned promise fulfills, it is fulfilled with the values from
...
+      }
+    }
+  });
+
+Returns a new L<Mojo::Promise> object that fulfills when all of the
+passed L<Mojo::Promise> objects have either fulfilled or rejected.
+
+It is fulfilled with the values from the fulfilled promises and reasons from
+rejected promises in the same order as the passed promises. This method can
+be useful for aggregating results of multiple promises.
```
Comment (kraih) [inline lib/Mojo/Promise.pm]: This needs a better description, and an example that explains the resulting data structure. Ideally inspired by the Mozilla docs.

File: lib/Mojo/Promise.pm [hunk 4]
```diff
@@ -288,6 +310,27 @@ rejects. If the returned promise fulfills, it is fulfilled with the values from
...
+  my $new = Mojo::Promise->all_settled(@promises);
+  $new->then(sub ($results) {
+    for my $result (@$results) {
+      if ($result->{status} eq 'fulfilled') {
+        say "fullfiled with $result->{value}";
+      } else {
+        say "rejected with $result->{reason}";
+      }
+    }
+  });
```
Comment (kraih) [inline lib/Mojo/Promise.pm]: This example is very inconsistent.

File: t/mojo/promise.t (hunk 2 above)
Comment (kraih) [reply t/mojo/promise.t]: And a test with mixed promises is also missing.

Comment (elcamlost) [reply t/mojo/promise.t]:
```
# All Settled some rejected
$promise  = Mojo::Promise->new->timeout(0.1)->resolve('first');
$promise2 = Mojo::Promise->new->timeout(0.2)->reject('second');
$promise3 = Mojo::Promise->new->timeout(0.3)->resolve('third');
@results  = ();
Mojo::Promise->all_settled($promise2, $promise3, $promise)
  ->then(sub { @results = @_ })->wait;
is_deeply(
  \@results,
  [
    {status => 'rejected',  reason => ['second']},
    {status => 'fulfilled', value  => ['third']},
    {status => 'fulfilled', value  => ['first']},
  ],
  'promises resolved'
);
```

What about this test?

File: t/mojo/promise.t [hunk 5]
```diff
@@ -213,6 +213,64 @@ Mojo::IOLoop->one_tick;
...
+$promise2->resolve('second');
+$promise3->resolve('third');
+$promise->resolve('first');
+Mojo::IOLoop->one_tick;
+is_deeply \@results,
+  [
+  {status => 'fulfilled', value => ['second']},
+  {status => 'fulfilled', value => ['first']},
+  {status => 'fulfilled', value => ['third']},
+  ],
```
[the remark this answers is not in the document]
Comment (elcamlost) [reply t/mojo/promise.t]: I couldn't find a way to preserve indentation and not to add brackets for is_deeply. So now all_settled tests call `is_deeply(...)` and other tests call `is_deeply ...`.
I'm not sure that it is ok, but see no other way with current perltidy rules.

File: t/mojo/promise.t [hunk 6]
```diff
@@ -213,6 +213,64 @@ Mojo::IOLoop->one_tick;
...
+$promise2->reject('second');
+$promise3->resolve('third');
+$promise->resolve('first');
+Mojo::IOLoop->one_tick;
+is_deeply \@results,
+  [
+  {status => 'rejected',  reason => ['second']},
+  {status => 'fulfilled', value  => ['first']},
+  {status => 'fulfilled', value  => ['third']},
+  ],
```
[the remark this answers is not in the document]
Comment (elcamlost) [reply t/mojo/promise.t]: Resolved, see [comment above](https://github.com/mojolicious/mojo/pull/1425#discussion_r336915737).

File: t/mojo/promise.t [hunk 7]
```diff
@@ -213,6 +213,64 @@ Mojo::IOLoop->one_tick;
...
+$promise2->reject('second');
+$promise3->reject('third');
+$promise->reject('first');
+Mojo::IOLoop->one_tick;
+is_deeply \@results,
+  [
+  {status => 'rejected', reason => ['second']},
+  {status => 'rejected', reason => ['first']},
+  {status => 'rejected', reason => ['third']},
+  ],
```
[the remark this answers is not in the document]
Comment (elcamlost) [reply t/mojo/promise.t]: Resolved, see [comment above](https://github.com/mojolicious/mojo/pull/1425#discussion_r336915737).

File: t/mojo/promise.t [hunk 8]
```diff
@@ -213,6 +213,64 @@ Mojo::IOLoop->one_tick;
...
+Mojo::IOLoop->one_tick;
+is_deeply \@results,
+  [
+  {status => 'rejected',  reason => ['second']},
+  {status => 'fulfilled', value  => ['first']},
+  {status => 'fulfilled', value  => ['third']},
+  ],
+  'promises resolved';
+
+# All Settled every rejected
```
[the remark this answers is not in the document]
Comment (elcamlost) [reply t/mojo/promise.t]: Changed to `# All Settled all rejected`

File: t/mojo/promise.t (hunk 1 above)
Comment (elcamlost) [reply t/mojo/promise.t]: fix

File: lib/Mojo/Promise.pm (hunk 4 above)
Comment (elcamlost) [reply lib/Mojo/Promise.pm]: Update example. 

I'm not sure that I was right to add this two lines of comments
```perl
# fulfilled: 3
# rejected: foo
```

I didn't find any example in mojolicious docs, when we `say` something in code snippet and also print the output in the comment to that snippet. But in this case
1. we have a more complex structure of the result
2. this style inspired by the MDN example for Promise.allSettled.

File: lib/Mojo/Promise.pm (hunk 3 above)
Comment (elcamlost) [reply lib/Mojo/Promise.pm]: Improve description.

File: t/mojo/promise.t (hunk 5 above)
[the remark this answers is not in the document]
Comment (kraih) [reply t/mojo/promise.t]: Of course there is, you can always use a variable like many if our other tests do.

File: t/mojo/promise.t [hunk 9]
```diff
@@ -213,6 +213,70 @@ Mojo::IOLoop->one_tick;
...
+  [
+    {status => 'fulfilled', value => ['second']},
+    {status => 'fulfilled', value => ['first']},
+    {status => 'fulfilled', value => ['third']},
+  ],
+  'promises fulfilled'
+);
+
+# All Settled some rejected
+# All Settled some rejected
```
Comment (kraih) [inline t/mojo/promise.t]: Double comment.

File: t/mojo/promise.t [hunk 10]
```diff
@@ -213,6 +213,70 @@ Mojo::IOLoop->one_tick;
...
+  \@results,
+  [
+    {status => 'rejected',  reason => ['second']},
+    {status => 'fulfilled', value  => ['third']},
+    {status => 'fulfilled', value  => ['first']},
+  ],
+  'promises mixed'
+);
+
+# All Settled all rejected
```
Comment (kraih) [inline t/mojo/promise.t]: That still doesn't sound quite right.

File: lib/Mojo/Promise.pm (hunk 4 above)
Comment (kraih) [reply lib/Mojo/Promise.pm]: Please follow the same style the rest of the documentation uses.

Comment (elcamlost) [reply lib/Mojo/Promise.pm]: Changed the comment to `# check the status of promises`

File: t/mojo/promise.t (hunk 9 above)
Comment (elcamlost) [reply t/mojo/promise.t]: Sorry, fixed it

File: t/mojo/promise.t (hunk 10 above)
Comment (kiwiroy) [reply t/mojo/promise.t]: Possibly include adverb, after/when. i.e. *All settled after some/all rejected*.  Might require slightly more verbosity though.

File: lib/Mojo/Promise.pm [hunk 11]
```diff
@@ -288,6 +310,40 @@ rejects. If the returned promise fulfills, it is fulfilled with the values from
...
+      say "$result->{status}: $result->{value}[0]"
+        if $result->{status} eq 'fulfilled';
+      say "$result->{status}: $result->{reason}[0]"
+        if $result->{status} eq 'rejected';
+    }
+  })->wait;
+
+Returns a new L<Mojo::Promise> object that fulfills when all of the
+passed L<Mojo::Promise> objects have either fulfilled or rejected.
+
```
Comment (kiwiroy) [inline lib/Mojo/Promise.pm]:
```pod
Returns a new L<Mojo::Promise> object that fulfills when all of the
supplied L<Mojo::Promise> objects have settled as I<fulfilled> or I<rejected>.
```

File: lib/Mojo/Promise.pm [hunk 12]
```diff
@@ -288,6 +310,40 @@ rejects. If the returned promise fulfills, it is fulfilled with the values from
...
+It is fulfilled with the list of hash references that each describes
+the outcome of each promise. Each hash has a C<status> key. If the C<status>
+is fulfilled, then a C<value> key is present. If the status is rejected,
+then a C<reason> key is present. The C<value> (or C<reason>) reflects what value
+each promise was fulfilled (or rejected) with.
+This method can be useful for aggregating results of multiple promises.
+Note that this method is B<EXPERIMENTAL> and might change without warning!
+
```
Comment (kiwiroy) [inline lib/Mojo/Promise.pm]:
```pod
The fulfillment handler is supplied with a list of hash references describing
the outcome of each promise. Each hash will have a C<status> key, with the
settlement state for the promise, and either a fulfillment C<value> or
rejection C<reason> key.
```

File: lib/Mojo/Promise.pm (hunk 4 above)
Comment (kraih) [reply lib/Mojo/Promise.pm]: There seems to be a misunderstanding, this is not about comments. The examples as a whole are unlike everything else in the Mojolicious docs.

Comment (elcamlost) [reply lib/Mojo/Promise.pm]: What exactly differs this example from other in mojolicious docs? Mojo::UserAgent has [similar examples](https://metacpan.org/release/Mojolicious/source/lib/Mojo/UserAgent.pm#L742).
It works, it demonstrates usage and output format of `all_settled` method.

I'd love to change example, but don't understand in what way should i do that.

**Comment by kraih** [comment]
The implementation seems mostly fine, but the docs and tests need to be redone completely. Since this is an already accepted feature, i tend towards closing this PR because it has gone stale. So it doesn't discourage others from opening new PRs.

**Comment by elcamlost** [comment]
It is stale, because I'm awaiting your answer to my questions. For example, [this](https://github.com/mojolicious/mojo/pull/1425#discussion_r337148297).

**Comment by kraih** [comment]
I see. If this PR depends on my participation, i'm afraid my limited time will prevent it from getting finished. Please feel free to look for help from others on IRC and the mailing-list, and try again in the future with a more complete PR.

File: lib/Mojo/Promise.pm (hunk 11 above)
Comment (karenetheridge) [reply lib/Mojo/Promise.pm]: is "fulfills" the preferred term here? I thought with promises, they are either *resolved* or rejected.

Comment (elcamlost) [reply lib/Mojo/Promise.pm]: Yes, It's part of the standard. See [mdn link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled).

**Pull request closed without merging**